test_workspace/recordset.bosatsu

```diff
@@ -1,37 +1,83 @@
+  REInt(value: w[Int])
+  REString(value: w[String])
+
+struct RecordField[t](name: String, to_entry: forall w. w[t] -> RowEntry[w])
+struct RecordValue[t](value: t)
+struct RecordGetter[shape, t](
+  field: shape[RecordField] -> RecordField[t],
+  value: shape[RecordValue] -> RecordValue[t]
+)
+struct RecordRowEntry[w, t](row_entry: RowEntry[w])
+
+struct RecordSet[shape](
+  fields: shape[RecordField],
+  rows: List[shape[RecordValue]],
+  getters: shape[RecordGetter[shape]],
+  traverse: forall w1,w2. shape[w1] -> (forall ss. w1[ss] -> w2[ss]) -> shape[w2],
+  record_to_list: forall w. shape[RecordRowEntry[w]] -> List[RowEntry[w]]
+)
 
-struct IntShape[string,int,bool](int: int)
-struct PairShape[shape,t,string,int,bool](left: t[string,int,bool], right: shape[string, int, bool])
-def make_pair_int_shape(rest: shape[string,int,bool], i: int) -> PairShape[shape,IntShape,string,int,bool]:
-  PairShape(IntShape(i), rest)
+def get(sh: shape[RecordValue], RecordGetter(_, getter): RecordGetter[shape, t]) -> t:
+  RecordValue(result) = sh.getter
+  result
+
+def create_field(rf: RecordField[t], fn: shape[RecordValue] -> t):
+  RecordGetter(\sh -> rf, \sh -> RecordValue(fn(sh)))
+
+def list_of_rows(RecordSet(fields, rows, getters, traverse, record_to_list): RecordSet[shape]):
+  def getter_to_row_entry(row: shape[RecordValue]):
+    (result_fn: forall tt. RecordGetter[shape, tt] -> RecordRowEntry[RecordValue, tt]) = \RecordGetter(get_field, get_value) ->
+      RecordField(name, to_entry) = get_field(fields)
+      RecordRowEntry(to_entry(get_value(row)))
+    result_fn
+  rows.map_List(\row -> record_to_list(getters.traverse(getter_to_row_entry(row))))
+
+struct RestructureOutput[shape1, shape2](
+  reshaperFields: shape1[RecordField] -> shape2[RecordField],
+  reshaperValues: shape1[RecordValue] -> shape2[RecordValue],
+  getters: shape2[RecordGetter[shape2]],
+  traverse: forall w1,w2. shape2[w1] -> (forall ss. w1[ss] -> w2[ss]) -> shape2[w2],
+  record_to_list: forall w. shape2[RecordRowEntry[w]] -> List[RowEntry[w]]
+)
+def restructure(RecordSet(fields, rows, getters, _, _): RecordSet[shape1], f: shape1[RecordGetter[shape1]] -> RestructureOutput[shape1, shape2]) -> RecordSet[shape2]:
+  RestructureOutput(reshaperF, reshaperV, new_getters, traverse, record_to_list) = f(getters)
+  RecordSet(reshaperF(fields), rows.map_List(reshaperV), new_getters, traverse, record_to_list)
 
-def add_int_field(RecordSet(fields, rows): RecordSet[shape1], name: String, f: shape1[String, Int, Bool] -> Int) -> RecordSet[PairShape[shape1, IntShape]]:
-  RecordSet(make_pair_int_shape(fields, Field(name, RVInt)), rows.map_List(\row -> make_pair_int_shape(row, f(row))))
+def concat_records(RecordSet(fields, rows, getters, traverse, record_to_list), more_rows):
+  RecordSet(fields, rows.concat(more_rows), getters, traverse, record_to_list)
 
-##################################################
+struct NilShape[w]
+struct PS[t,rest,w](left: w[t], right: rest[w])
```

Comment:
dude, you are a type maniac.... these are pretty wild types.... certainly stress testing things here. :)

Reply:
Honestly, I think this is the first time I've ever felt I've understood a type system enough to be creative with it.